### PR TITLE
Introduce new APIs into Tiger to manage CSP directive

### PR DESCRIPTION
**Pull request by dlcd2203**
JIRA: SD-2456

---

Supported PR commands:

| Command                  | Description             |
| ------------------------ | ----------------------- |
| `ok to test`             | Re-run standard checks  |
| `extended test`          | BackstopJS tests        |
| `extended check sonar`   | SonarQube tests         |
| `extended check cypress` | Cypress E2E tests       |
| `extended check plugins` | Dashboard plugins tests |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `check-extended-cypress` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/10030/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/11997/ : SUCCESS in 13m 19s

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 1]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
             });
         });
     },
+    getAllCSPDirectives: async (): Promise<Array<ICSPDirective>> => {
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1070]: Please add an empty line to separate this function from the one above.

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 2]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
...
+        try {
+            return await authApiCall(async (sdk) => {
+                const result = await sdk.entities.getAllEntitiesCspDirectives({});
+                return result.data?.data || [];
+            });
+        } catch (error) {
+            return [];
+        }
+    },
+    getCSPDirective: async (directiveId: string): Promise<ICSPDirective> => {
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1080]: Please add an empty line to separate this function from the one above.

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 3]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
...
+        try {
+            return await authApiCall(async (sdk) => {
+                const result = await sdk.entities.getEntityCspDirectives({ id: directiveId });
+                return result.data?.data;
+            });
+        } catch (error: any) {
+            throw convertApiError(error);
+        }
+    },
+    createCSPDirective: async (requestData: ICSPDirective): Promise<ICSPDirective> => {
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1090]: Please add an empty line to separate this function from the one above.

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 4]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
...
+            if (error.response?.status === 400) {
+                const message = error?.response?.data?.detail
+                    ? error?.response?.data?.detail
+                    : "Server error";
+                throw new UnexpectedError(message, error);
+            }
+            throw convertApiError(error);
+        }
+    },
+    updateCSPDirective: async (directiveId: string, requestData: ICSPDirective): Promise<ICSPDirective> => {
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1116]: Please add an empty line to separate this function from the one above.

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 5]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
...
+            if (error.response?.status === 400) {
+                const message = error?.response?.data?.detail
+                    ? error?.response?.data?.detail
+                    : "Server error";
+                throw new UnexpectedError(message, error);
+            }
+            throw convertApiError(error);
+        }
+    },
+    deleteCSPDirective: async (directiveId: string) => {
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1143]: Please add an empty line to separate this function from the one above.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/10031/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/11998/ : SUCCESS in 11m 00s

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 6]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
...
         });
     },
+    getAllCSPDirectives: async (): Promise<Array<ICSPDirective>> => {
+        try {
+            return await authApiCall(async (sdk) => {
+                const result = await sdk.entities.getAllEntitiesCspDirectives({});
+                return result.data?.data || [];
+            });
+        } catch (error) {
+            return [];
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1077]: Should we call `throw convertApiError(error);` to throw an error up to the app? I think the SDK should not hide any error at all.

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 7]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
...
+                return result.data?.data;
+            });
+        } catch (error: any) {
+            if (error.response?.status === 400) {
+                const message = error?.response?.data?.detail
+                    ? error?.response?.data?.detail
+                    : "Server error";
+                throw new UnexpectedError(message, error);
+            }
+            throw convertApiError(error);
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1113]: Can we delete lines 1107 ~ 1112 and keep just this line only? Logic from lines 1107 ~ 1112 is probably already contained in `convertApiError`...

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 8]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
...
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1140]: Can we delete lines 1134 ~ 1139 and keep just this line only? Logic from lines 1134 ~ 1139 is probably already contained in `convertApiError`...

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 9]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
...
+                const result = await sdk.entities.getEntityCspDirectives({ id: directiveId });
+                return result.data?.data;
+            });
+        } catch (error: any) {
+            throw convertApiError(error);
+        }
+    },
+    createCSPDirective: async (requestData: ICSPDirective): Promise<ICSPDirective> => {
+        try {
+            const jsonApiCspDirectiveInDocument: JsonApiCspDirectiveInDocument = {
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1092]: Please move lines 1092 ~ 1098 into the callback function (that gets passed into `authApiCall` as a parameter) at line 1101. We only need to declare `jsonApiCspDirectiveInDocument` there, don't we?

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 10]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
...
+                    ? error?.response?.data?.detail
+                    : "Server error";
+                throw new UnexpectedError(message, error);
+            }
+            throw convertApiError(error);
+        }
+    },
+    updateCSPDirective: async (directiveId: string, requestData: ICSPDirective): Promise<ICSPDirective> => {
+        try {
+            const jsonApiCspDirectiveInDocument: JsonApiCspDirectiveInDocument = {
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1118]: Please move lines 1118 ~ 1124 into the callback function (that gets passed into `authApiCall` as a parameter) at line 1127. We only need to declare `jsonApiCspDirectiveInDocument` there, don't we?

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 1 above)
Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1070]: not sure about this. Almost functions do not have an empty line to separate each one

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 2 above)
Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1080]: not sure about this. Almost functions do not have an empty line to separate each one

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 3 above)
Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1090]: not sure about this. Almost functions do not have an empty line to separate each one

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 5 above)
Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1143]: not sure about this. Almost functions do not have an empty line to separate each one

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 1 above)
Comment (anvuong119) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1070]: ok then

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 2 above)
Comment (anvuong119) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1080]: ok then

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 5 above)
Comment (anvuong119) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1143]: ok then

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 7 above)
Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1113]: `convertApiError` did not handle the statusCode = 400 case.

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 8 above)
Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1140]: `convertApiError` did not handle the statusCode = 400 case.

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 9 above)
Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts]: fixed

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 10 above)
Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts]: fixed

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 6 above)
Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1077]: It's the same https://github.com/gooddata/gooddata-ui-sdk/blob/9b589cc7118952b6540d753e49a96e7b69cdd655/libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts#L458

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/10061/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/12033/ : SUCCESS in 10m 21s

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 8 above)
Comment (anvuong119) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1140]: It does return `UnexpectedError` at the following place: https://github.com/gooddata/gooddata-ui-sdk/blob/2fac540cb6cae33df4239e9e92cc97e3d949bcda/libs/sdk-backend-tiger/src/utils/errorHandling.ts#L35

→ Do we really need to get `message` from the error and then pass it to UnexpectedError's constructor?

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 7 above)
Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1113]: Screen Shot 2022-10-19 at 16 11 52
So we need to handle  specific bad request case to get detailed message

Comment (anvuong119) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1113]: ok then

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 11]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
...
+                    jsonApiCspDirectiveInDocument,
+                });
+
+                return result.data?.data;
+            });
+        } catch (error: any) {
+            if (error.response?.status === 400) {
+                const message = error?.response?.data?.detail
+                    ? error?.response?.data?.detail
+                    : "Server error";
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1110]: Status code 400 means Bad request, doesn't it?

→ Using `"Server error"` here seems to be unsuitable. Please change it to `"An unexpected error has occurred"` or `"Bad request"` instead.

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 12]
```diff
@@ -1050,4 +1067,86 @@ export const buildTigerSpecificFunctions = (
...
```
Comment (anvuong119) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1137]: Status code 400 means Bad request, doesn't it?

→ Using `"Server error"` here seems to be unsuitable. Please change it to `"An unexpected error has occurred"` or `"Bad request"` instead.

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts (hunk 11 above)
Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:1110]: I have cloned the code at here
https://github.com/gooddata/gooddata-ui-sdk/blob/9b589cc7118952b6540d753e49a96e7b69cdd655/libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts#L511

**Comment by anvuong119** [comment]
LGTM

**Comment by dlcd2203** [comment]
extended check sonar

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-check-extended-sonar-PR3066

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/10094/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/2064/ : SUCCESS in 18m 59s

File: libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts [hunk 13]
```diff
@@ -202,6 +204,16 @@ export interface IInvitationUserResponse {
     errorMessage?: string;
 }
 
+/**
+ * @internal
+ */
+export interface ICSPDirective {
+    id: string;
+    attributes: {
+        sources: Array<string>;
```
Comment (kandl) [inline libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:213]: Please could we add some docs? At least, what is this array of strings, as string can be almost anything...

Comment (dlcd2203) [reply libs/sdk-backend-tiger/src/backend/tigerSpecificFunctions.ts:217]: Added

**Review by kandl — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by dlcd2203** [comment]
Rebase master

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/10122/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/12105/ : SUCCESS in 10m 06s

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR3066

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/2922/ : FAILURE in 3m 57s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/3228/ : SUCCESS in 30s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/2070/ : SUCCESS in 22m 53s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/12106/ : SUCCESS in 10m 50s

**Pull request merged**